auth/standard_role_manager.cc

```diff
@@ -713,4 +714,135 @@ future<> standard_role_manager::remove_attribute(std::string_view role_name, std
                 {sstring(role_name), sstring(attribute_name)});
     }
 }
+
+future<std::vector<role_info>> standard_role_manager::get_role_info(const bool with_salted_hashes) const {
+    // This is super verbose, but it's difficult to avoid it.
+    // Unfortunately, fmt isn't able to produce `constexpr std::string_view`s.
+    const auto role_query = std::invoke([&] {
+        if (with_salted_hashes) {
+            return seastar::format("SELECT {}, {}, {}, {} FROM {}.{}",
+                meta::roles_table::role_col_name,
+                meta::roles_table::can_login_col_name,
+                meta::roles_table::is_superuser_col_name,
+                meta::roles_table::salted_hash_col_name,
+                get_auth_ks_name(_qp),
+                meta::roles_table::name);
+        }
+
+        return seastar::format("SELECT {}, {}, {} FROM {}.{}",
+                meta::roles_table::role_col_name,
+                meta::roles_table::can_login_col_name,
+                meta::roles_table::is_superuser_col_name,
+                get_auth_ks_name(_qp),
+                meta::roles_table::name);
+    });
+
+    const auto roles = co_await _qp.execute_internal(
+            role_query,
+            db::consistency_level::LOCAL_ONE, // TODO: Verify later if this is okay.
+            internal_distributed_query_state(),
+            cql3::query_processor::cache_internal::yes);
+
+    if (roles->empty()) [[unlikely]] {
+        on_internal_error(log, "No role has been found in the system");
+    }
+
+    std::vector<role_info> result{};
+    result.reserve(roles->size());
+
+    for (const auto& role_record : *roles) {
+        role_info ri{};
+
+        ri.role_name = role_record.get_as<sstring>(meta::roles_table::role_col_name);
+        ri.config.can_login = role_record.get_or<bool>(meta::roles_table::can_login_col_name, false);
+        ri.config.is_superuser = role_record.get_or<bool>(meta::roles_table::is_superuser_col_name, false);
+
+        if (with_salted_hashes && role_record.has(meta::roles_table::salted_hash_col_name)) {
+            ri.maybe_salted_hash = role_record.get_as<sstring>(meta::roles_table::salted_hash_col_name);
+        }
+
+        result.push_back(std::move(ri));
+        co_await coroutine::maybe_yield();
+    }
+
+    co_return result;
+}
+
+future<std::vector<role_grants>> standard_role_manager::get_role_grants() const {
+    // We don't use the `role_members` table here because it keeps track of this data
+    // in a "dense" way, i.e. pairs (granted role, grantee), so we might end up browsing
+    // O(n^2) records where n is the number of roles. Using the `roles` table, we are
+    // guaranteed to get at most O(n) rows.
```

Comment:
Are you sure about this? `role_grants` seems to represents just the grants that were assigned by the user, not the transitive closure. I checked it:

```cql
cassandra@cqlsh> create role ua;
cassandra@cqlsh> create role ub;
cassandra@cqlsh> create role uc;
cassandra@cqlsh> grant ua to ub;
cassandra@cqlsh> grant ub to uc;
cassandra@cqlsh> select * from system.role_members;

 role | member
------+--------
   ub |     uc
   ua |     ub

(2 rows)
```

If it were as you wrote I would expect there to be one more row: `ua, uc`.

Reply:
By the way, we already have `standard_role_manager::query_all_directly_granted()` which seems to be doing it via `role_members`.